src/ess/imaging/io.py

```diff
@@ -0,0 +1,358 @@
+    *,
+    file_path: FilePath,
+    image_detector_name: ImageDetectorName,
+    histogram_mode_detectors_path: HistogramModeDetectorsPath = DEFAULT_HISTOGRAM_PATH,
+) -> HistogramModeDetector:
+    with snx.File(file_path, mode="r") as f:
+        img_path = f"{histogram_mode_detectors_path}/{image_detector_name}"
+        dg: sc.DataGroup = f[img_path][()]
+
+    # Manually assign unit to the histogram detector mode data
+    img: sc.DataArray = dg['data']
+    if (original_unit := img.unit) != 'counts':
+        img.unit = 'counts'
+        warnings.warn(
+            f"The unit of the histogram detector data is [{original_unit}]. "
+            f"It is expected to be [{img.unit}]. "
+            f"The loader manually assigned the unit to be [{img.unit}].",
+            stacklevel=0,
+        )
+    return HistogramModeDetector(dg)
+
+
+MinDim1 = NewType("MinDim1", sc.Variable | None)
+"""Minimum value of the first dimension."""
+MaxDim1 = NewType("MaxDim1", sc.Variable | None)
+"""Maximum value of the first dimension."""
+MinDim2 = NewType("MinDim2", sc.Variable | None)
+"""Minimum value of the second dimension."""
+MaxDim2 = NewType("MaxDim2", sc.Variable | None)
+
+
+def _make_coord_if_needed(da: sc.DataArray, dim: str) -> None:
+    if dim not in da.coords.keys():
+        da.coords[dim] = sc.arange(dim=dim, start=0, stop=da.sizes[dim])
+
+
+def separate_detector_images(
+    dg: HistogramModeDetector,
+    min_dim_1: MinDim1,
+    max_dim_1: MaxDim1,
+    min_dim_2: MinDim2,
+    max_dim_2: MaxDim2,
+) -> HistogramModeDetectorData:
+    da: sc.DataArray = sc.sort(dg['data'], 'time')
+    # Assign position coordinates to the detector data
+    _make_coord_if_needed(da, DIM1_COORD_NAME)
+    _make_coord_if_needed(da, DIM2_COORD_NAME)
+    # Crop the detector data by the given coordinates
+    da = da[DIM1_COORD_NAME, min_dim_1:max_dim_1][DIM2_COORD_NAME, min_dim_2:max_dim_2]
+    return HistogramModeDetectorData(da)
+
+
+def separate_image_key_logs(*, dg: HistogramModeDetector) -> ImageKeyLogs:
+    return ImageKeyLogs(sc.sort(dg['image_key']['value'], key='time'))
+
+
+def load_nexus_rotation_logs(
+    file_path: FilePath,
+    motion_sensor_name: RotationMotionSensorName,
+) -> RotationLogs:
+    log_path = f"entry/instrument/{motion_sensor_name}/rotation_stage_readback"
+    with snx.File(file_path, mode="r") as f:
+        return RotationLogs(f[log_path][()]['value'])
+
+
+def derive_log_coord_by_range(
+    da: sc.DataArray, log: sc.DataArray, out_of_range_value: sc.Variable
+) -> sc.Variable:
+    """Sort the logs by time and decide which log entry corresponds to each time bin.
+
+    It assumes a log value is valid until the next log entry.
+    """
+    sorted_logs = sc.sort(log, 'time')
+    min_time = sc.datetime(da.coords['time'].min().value - 1, unit='ns')
+    max_time = sc.datetime(da.coords['time'].max().value + 1, unit='ns')
+    time_edges = sc.concat((min_time, sorted_logs.coords['time'], max_time), 'time')
```

Comment:
It was because of the out-or-range-values and assumming that logs are always less than the actual time coordinate but I think it can go wrong pretty easily... 

I updated it so that it just slice the dataset according to the log.